### PR TITLE
Update scaled_dot_product_attention scaling factor from our derivation.

### PR DESCRIPTION
**Pull request by DouglasOrr**
It was derived by assuming a fixed normaliser for the softmax, such that the output is a LogNormal distribution that is normalised in expectation (rather than explicitly over a finite input). This gives a reasonable rule for temperature=1 and long sequences >= 64. It is not drastically different from the old rule, the differences mainly appearing for longer sequences (e.g. 4096).

`torch.nn.functional`

image

`unit_scaling.functional` (old)

image

**`unit_scaling.functional` (new)**

image

---

For my/future reference.

```py
class TestM(nn.Module):
    def forward(self, q: Tensor, k: Tensor, v: Tensor) -> Tensor:
        # return F.scaled_dot_product_attention(q, k, v)
        return U.scaled_dot_product_attention(q, k, v)

def run(d: int, n: int, name: str) -> Dict[str, Tensor]:
    q, k, v = (x.requires_grad_() for x in torch.randn(3, 100, n, d))
    y = TestM()(q, k, v)
    y.backward(torch.randn_like(y))
    return dict(name=name, d=d, n=n, y=float(y.std()), grad_q=float(q.grad.std()), grad_k=float(k.grad.std()), grad_v=float(v.grad.std()))

pd.DataFrame.from_dict([
    run(name="baseline", d=64, n=256),
    run(name="vary d", d=16, n=256),
    run(name="vary d", d=256, n=256),
    run(name="vary n", d=64, n=64),
    run(name="vary n", d=64, n=1024),
    run(name="vary n", d=64, n=4096),
]).style.format({c:"{:.2f}" for c in ["y", "grad_q", "grad_k", "grad_v"]}).hide()
```

### CODE REVIEW

**Review by thecharlieblake — APPROVED** [review]
Thanks so much for doing this. Much much happier with this version!

(good spot with the dropout too!)
